src/session-user/session-user.service.ts

```diff
@@ -1,4 +1,117 @@
+import { GetUserDto } from 'src/user/dtos/get-user.dto';
+import { UserService } from 'src/user/user.service';
+import { STORYBLOK_STORY_STATUS_ENUM } from 'src/utils/constants';
+import { CourseUserService } from '../course-user/course-user.service';
+import { CourseService } from '../course/course.service';
+import { IFirebaseUser } from '../firebase/firebase-user.interface';
+import { CreateSessionUserDto } from './dtos/create-session-user.dto';
+import { SessionUserRepository } from './session-user.repository';
 
 @Injectable()
-export class SessionUserService {}
+export class SessionUserService {
+  constructor(
+    @InjectRepository(SessionUserRepository) private sessionUserRepository: SessionUserRepository,
+    private readonly courseUserService: CourseUserService,
+    private readonly userService: UserService,
+    private readonly sessionService: SessionService,
+    private readonly courseService: CourseService,
+  ) {}
+
+  private markCourseComplete(userCourse: GetUserDto, courseSession: SessionEntity[]): boolean {
+    const userCompletedSession = userCourse.course[0].session;
+
+    const userSessionIds = userCompletedSession.map((session) => {
+      if (session.completed) return session.id;
+    });
+
+    const courseSessionIds = courseSession.map((session) => {
+      if (session.status === STORYBLOK_STORY_STATUS_ENUM.PUBLISHED) return session.id;
+    });
+
+    return _.xor(courseSessionIds, userSessionIds).length == 0;
+  }
+
+  public async createSessionUser(
+    firebaseUser: IFirebaseUser,
+    { sessionId }: CreateSessionUserDto,
+  ): Promise<SessionUserEntity> {
+    const { courseId } = await this.sessionService.getCourseFromSessionId(sessionId);
+
+    const courseSessions = await this.courseService.getCourseSessions(courseId);
+
+    if (!courseSessions) {
+      throw new HttpException('COURSE SESSIONS NOT FOUND', HttpStatus.NOT_FOUND);
+    }
+
+    const { id } = await this.userService.getUserFromFirebaseUid(firebaseUser);
+
+    let courseUser = await this.courseUserService.courseUserExists({ userId: id, courseId });
+
+    if (!courseUser) {
+      courseUser = await this.courseUserService.createCourseUser({ userId: id, courseId });
+    }
+
+    const createSessionUserObject = this.sessionUserRepository.create({
+      sessionId,
+      courseUserId: courseUser.id,
+      completed: false,
+    });
+
+    let sessionUser = await this.sessionUserRepository.findOne({
+      courseUserId: courseUser.id,
+      sessionId,
+    });
+
+    if (!sessionUser) {
+      sessionUser = await this.sessionUserRepository.save(createSessionUserObject);
+    }
+
+    return sessionUser;
+  }
+
+  public async updateSessionUser(firebaseUser: IFirebaseUser, sessionId: string) {
+    const { courseId } = await this.sessionService.getCourseFromSessionId(sessionId);
+
+    const courseSessions = await this.courseService.getCourseSessions(courseId);
+
+    if (!courseSessions) {
```

Comment:
this condition `!courseSessions` would never happen right? If the user is interacting with a session (meaning its published), and there is a `courseId` then the course definitely has that session at least. So i think this check can be removed.

It might be worth checking for `!courseId` instead, in case theres no session in the db for the provided `sessionId` - should be an edge case if our storyblok webhook failed, but possible!